### PR TITLE
Bluetooth: Audio: Broadcast Assistant: Check add_source duplication

### LINKED ISSUE #69661: Bluetooth: LE Audio: Broadcast Assistant and Scan Delegator: Reject Add Source operations that would result in duplicates
**Is your enhancement proposal related to a problem? Please describe.**
According to BAP 6.5.4. the Broadcast Assistant is not allowed to add sources that would result in duplications. We must therefor add functionality in the Broadcast Assistant such that it does not send a Add Source operation request for a source that is already present.

Although not (currently) a requirement in the Scan Delegator, the Scan Delegator should check for duplicates as well, otherwise it is fully possible that it will create duplicates (which is the opposite of the intent). E.g. consider a Broadcast Assistant that legally sends an Add Source Operation, but before it has arrived the same Broadcast Source was added in another manner. Even though the Assistant made sure it would not create a duplicate, that is now the side-effect.

**Describe the solution you'd like**
The Broadcast Source always keeps track of the sources that has been added, mirroring the recv_states of the Scan Delegator. If it receives an error on the Add Operation it needs to read out all receives states to update its mirror. Only reading on failure would prevent many operations, which could be undesirable during heavy workload, e.g. when one assistant is coupled to many sinks and wants to add source to them all.

Scan Delegator needs to check the source being added and guard against duplicates, and send error when applicable.

**Describe alternatives you've considered**
Before sending an Add Source operation, the assistant has to read all of the recv_states from the Scan Delegator and make sure it does not send a duplicate.

Scan Delegator needs to check the source being added and guard against duplicates, and send error when applicable.

### PR DESCRIPTION
**Pull request by fredrikdanebjer**
Added a check for checking duplication when adding a source in the broadcast assistant, as per requirement in BAP 6.5.4. It is not permitted to add a source such that a combination of broadcast_id, adv_sid and address type would result in same triple of values. Added a recv_state struct to cache these values when read, and moved past_avail and srd_ids into this as well.

Since there is no application specific error code present today, the assistant would not know if an add_src failed due to duplication, and so this has not been added to this PR. We should get an errata fix for this first.

This partially fixes https://github.com/zephyrproject-rtos/zephyr/issues/69661

### CODE REVIEW

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 1]
```diff
@@ -1013,6 +1029,21 @@ int bt_bap_broadcast_assistant_add_src(struct bt_conn *conn,
...
+		if (param->addr.type == state->addr.type && param->adv_sid == state->adv_sid &&
+		    param->broadcast_id == state->broadcast_id) {
+			LOG_DBG("recv_state using broadcast_id=0x%06X, adv_sid=0x%02X, and "
+				"addr.type=0x%02X already exists with src_id=0x%02X",
+				param->broadcast_id, param->adv_sid, param->addr.type,
+				state->src_id);
+
+			return -EINVAL;
+		}
+	}
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:1045]: Similarly to the BASS update, suggest to move this into a function. 

May want to consider finding a way to reuse the same function

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 2]
```diff
@@ -782,8 +796,9 @@ static int broadcast_assistant_reset(struct bap_broadcast_assistant_instance *in
 	(void)k_work_cancel_delayable(&inst->bap_read_work);
 
 	for (int i = 0U; i < CONFIG_BT_BAP_BROADCAST_ASSISTANT_RECV_STATE_COUNT; i++) {
-		inst->src_ids[i] = 0U;
-		inst->past_avail[i] = false;
+		memset(&inst->recv_states[i], 0,
+		       sizeof(struct bap_broadcast_assistant_recv_state_info));
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:800]:
```suggestion
		memset(&inst->recv_states[i], 0, sizeof(inst->recv_states[i]);
```

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 3]
```diff
@@ -34,17 +35,21 @@ LOG_MODULE_REGISTER(bt_bap_broadcast_assistant, CONFIG_BT_BAP_BROADCAST_ASSISTAN
...
 #define MINIMUM_RECV_STATE_LEN          15
 
+struct bap_broadcast_assistant_recv_state_info {
+	uint8_t src_id;
+	bt_addr_le_t addr;
+	uint8_t adv_sid;
+	uint32_t broadcast_id;
+
+	/** Cached PAST available */
+	bool past_avail;
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:45]:
```suggestion
	uint8_t src_id;

	/** Cached PAST available */
	bool past_avail;
	uint8_t adv_sid;
	uint32_t broadcast_id;
	bt_addr_le_t addr;
```

Suggest a reordering to put the 3 1-octet values in the same word

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 1 above)
Comment (niym-ot) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c]: Done

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 2 above)
Comment (niym-ot) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c]: Done

File: subsys/bluetooth/audio/bap_broadcast_assistant.c (hunk 3 above)
Comment (niym-ot) [reply subsys/bluetooth/audio/bap_broadcast_assistant.c]: Done

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 4]
```diff
@@ -781,6 +780,28 @@ static struct bt_le_scan_cb scan_cb = {
...
+
+		if (state != NULL && state->broadcast_id == broadcast_id &&
+			state->adv_sid == adv_sid && state->addr.type == addr_type) {
+			LOG_DBG("recv_state already exists at src_id=0x%02X", state->src_id);
+
+			return true;
+		}
+	}
+	return false;
+}
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:803]:
```suggestion
	}
	
	return false;
}
```

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 5]
```diff
@@ -781,6 +780,28 @@ static struct bt_le_scan_cb scan_cb = {
...
+ * if the operation would result in duplicate values for the combined Source_Address_Type,
+ * Source_Adv_SID, and Broadcast_ID fields of any Broadcast Receive State characteristic exposed
+ * by the Scan Delegator.
+ */
+static bool broadcast_src_is_duplicate(uint32_t broadcast_id, uint8_t adv_sid, uint8_t addr_type)
+{
+	struct bap_broadcast_assistant_recv_state_info *state;
+
+	for (size_t i = 0; i < ARRAY_SIZE(broadcast_assistant.recv_states); i++) {
+		state = &broadcast_assistant.recv_states[i];
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:793]:
```suggestion
	for (size_t i = 0; i < ARRAY_SIZE(broadcast_assistant.recv_states); i++) {
		const struct bap_broadcast_assistant_recv_state_info *state = &broadcast_assistant.recv_states[i];
```

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 6]
```diff
@@ -1030,20 +1049,11 @@ int bt_bap_broadcast_assistant_add_src(struct bt_conn *conn,
 	}
 
 	/* Check if this operation would result in a duplicate before proceeding */
-	for (size_t i = 0; i < ARRAY_SIZE(broadcast_assistant.recv_states); i++) {
-		state = &broadcast_assistant.recv_states[i];
+	if (broadcast_src_is_duplicate(param->broadcast_id, param->adv_sid, param->addr.type)) {
+		LOG_DBG("source already exists");
```
Comment (Thalley) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:1053]:
```suggestion
		LOG_DBG("Broadcast source already exists");
```

**Review by Thalley — CHANGES_REQUESTED** [review]
Minor requests, otherwise LGTM

**Review by Thalley — APPROVED** [review]

**Review by larsgk — APPROVED** [review]

**Pull request merged**